starfish/image/_filter/__init__.py

```diff
@@ -19,38 +19,30 @@
 
 class Filter(PipelineComponent):
 
-    filter_group: argparse.ArgumentParser
-
     @classmethod
     def _get_algorithm_base_class(cls) -> Type[AlgorithmBase]:
         return _base.FilterAlgorithmBase
 
     @classmethod
-    def _add_to_parser(cls, subparsers):
-        """Adds the filter component to the CLI argument parser."""
-        filter_group = subparsers.add_parser("filter")
-        filter_group.add_argument("-i", "--input", type=FsExistsType(), required=True)
-        filter_group.add_argument("-o", "--output", required=True)
-        filter_group.set_defaults(starfish_command=Filter._cli)
-        filter_subparsers = filter_group.add_subparsers(dest="filter_algorithm_class")
-
-        for algorithm_cls in cls._algorithm_to_class_map().values():
-            group_parser = filter_subparsers.add_parser(algorithm_cls._get_algorithm_name())
-            group_parser.set_defaults(filter_algorithm_class=algorithm_cls)
-            algorithm_cls._add_arguments(group_parser)
-
-        cls.filter_group = filter_group
-
-    @classmethod
-    def _cli(cls, args, print_help=False):
-        """Runs the filter component based on parsed arguments."""
-
-        if args.filter_algorithm_class is None or print_help:
-            cls.filter_group.print_help()
-            cls.filter_group.exit(status=2)
-
-        print('Filtering images ...')
-        stack = ImageStack.from_path_or_url(args.input)
-        instance = args.filter_algorithm_class(**vars(args))
-        output = instance.run(stack)
-        output.write(args.output)
+    def _cli_run(cls, ctx, instance):
+        output = ctx.obj["output"]
+        stack = ctx.obj["stack"]
+        filtered = instance.run(stack)
+        filtered.write(output)
+
+@click.group("filter")
+@click.option("-i", "--input", type=click.Path(exists=True))
+@click.option("-o", "--output", required=True)
+@click.pass_context
+def _cli(ctx, input, output):
+    print("Filtering images...")
+    ctx.obj = dict(
+        component=Filter,
+        input=input,
+        output=output,
+        stack=ImageStack.from_path_or_url(input),
+    )
+
+
+Filter._cli = _cli  # type: ignore
```

Comment:
This (and the register line below) need to live outside the class because we need this to be executed when the module is imported?

Reply:
This line in particular is mostly stemming from the fact that `@classmethod` and `@click.command` didn't work together (see https://github.com/spacetx/starfish/pull/722/commits/e591f867678d2ee19a3f4f5ec715f8b1f0e41903) But yes, the registration must happen at import time currently, unless, e.g., each module provides a `register` method that `starfish.starfish` would then call.

Reply:
I could have sworn I tried with `@staticmethod` and that it didn't work, but perhaps there was another issue that I've since fixed. @ttung's suggestion certainly looks much better.